### PR TITLE
Optimization: do not add control qubits to compute/uncompute steps of compute_action idiom.

### PR DESCRIPTION
**Pull request by schweitzpgi**
Apply control to kernel with a compute_action.

In order to be somewhat agnostic to phase ordering, the pass will recognize both the high-level quake.compute_action op and the triple of quake.apply operations that quake.compute_action may be expanded to. Regardless of the presentation in the IR, the optimization is to omit the control from the compute and uncompute steps and only add it to the action step.

Add a test for both apply specialization and apply specialization with lambda lifting.

### CODE REVIEW

File: lib/Optimizer/Transforms/ApplyOpSpecialization.cpp [hunk 1]
```diff
@@ -337,40 +376,88 @@ class ApplySpecializationPass
...
+    auto loc = func.getLoc();
     SmallVector<Type> inTys = {qvecTy};
     inTys.append(funcTy.getInputs().begin(), funcTy.getInputs().end());
     auto newFunc = cudaq::opt::factory::createFunction(
         funcName, funcTy.getResults(), inTys, module);
     newFunc.setPrivate();
     IRMapping mapping;
-    funcOp.getBody().cloneInto(&newFunc.getBody(), mapping);
+    func.getBody().cloneInto(&newFunc.getBody(), mapping);
+    auto controlNotNeeded = computeActionAnalysis(newFunc);
```
Comment (amccaskey) [inline lib/Optimizer/Transforms/ApplyOpSpecialization.cpp:449]: Is it possible to setup a flag or something such that we can control whether we do this or not? The flag should always default to true, and it is probably best that it is not something we document or talk too much about (there's no reason real users will need this). My objective here is to provide a way to demonstrate key features about the compiler - plots that show how the code generated here is optimal and leads to a decrease in the number of controlled operations required. We may also want this for testing with CircuitCheck (make sure the unitaries are the same if we do this and if we don't).

File: test/AST-Quake/compute_action-2.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+// RUN: cudaq-quake %s | cudaq-opt --canonicalize --apply-op-specialization | FileCheck %s
```
Comment (amccaskey) [inline test/AST-Quake/compute_action-2.cpp:9]: Great test.

**Comment by amccaskey** [comment]
Overall LGTM. Just a few questions and possible follow-ups.

File: lib/Optimizer/Transforms/ApplyOpSpecialization.cpp [hunk 3]
```diff
@@ -243,13 +281,14 @@ static bool hasMonotonicControlInduction(cudaq::cc::LoopOp loop) {
...
-// loop such as `for(i = 0; i < n; ++i)` is a counted loop that must execute
+// A monotonic loop is defined to be a loop that will execute some bounded
+// number of iterations that can be predetermined before the loop, in fact,
+// executes. A loop such as `for(i = 0; i < n; ++i)` is a monotonic loop that
+// must execute
 //   n : if n > 0
 //   0 : if n <= 0
 // iterations. Early exits (break statements) are not permitted.
-static bool isaCountedLoop(Operation &op) {
+static bool isaMonotonicLoop(Operation &op) {
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Transforms/ApplyOpSpecialization.cpp:291]: No. I had switched gears in an earlier version of this code and didn't update the name. Here I updated the name to be consistent with the rest of the code. (This is finding loops that step by some value _i_ .)

File: lib/Optimizer/Transforms/ApplyOpSpecialization.cpp (hunk 1 above)
Comment (amccaskey) [reply lib/Optimizer/Transforms/ApplyOpSpecialization.cpp:449]: it's ok to not be "truly" hidden.

Comment (schweitzpgi) [reply lib/Optimizer/Transforms/ApplyOpSpecialization.cpp:455]: Done.

**Review by amccaskey — APPROVED** [review]

**Pull request merged**